### PR TITLE
Add Apache flink example.

### PR DESCRIPTION
**Pull request by mindprince**
We have this config running in production. We use Flink 1.1.4

### CODE REVIEW

File: example_configs/flink.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      job: $2
+
+  ## Task Manager ##
+  # Example: org.apache.flink.metrics<key0=debian-8, key1=taskmanager, key2=407401ecfef1f050d3dcc83af9cbfb74, key3=My_Job, key4=Map, key6=1, name=numRecordsIn><>Count
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=taskmanager, key2=(.*), key3=(.*), key4=(.*), key5=([0-9]*), name=([a-z].*)><>Count
+    name: flink_taskmanager_$6
+    labels:
+      host: $1
+      taskmanager_id: $2
+      job: $3
```
Comment (brian-brazil) [inline example_configs/flink.yml]: job has a meaning already in Prometheus, this needs another name

File: example_configs/flink.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      ip: $1
+      job: $2
+
+  ## Task Manager ##
+  # Example: org.apache.flink.metrics<key0=debian-8, key1=taskmanager, key2=407401ecfef1f050d3dcc83af9cbfb74, key3=My_Job, key4=Map, key6=1, name=numRecordsIn><>Count
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=taskmanager, key2=(.*), key3=(.*), key4=(.*), key5=([0-9]*), name=([a-z].*)><>Count
+    name: flink_taskmanager_$6
+    labels:
+      host: $1
+      taskmanager_id: $2
```
Comment (brian-brazil) [inline example_configs/flink.yml]: What's the cardinality of this like?

File: example_configs/flink.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    labels:
+      ip: $1
+      job: $2
+
+  ## Task Manager ##
+  # Example: org.apache.flink.metrics<key0=debian-8, key1=taskmanager, key2=407401ecfef1f050d3dcc83af9cbfb74, key3=My_Job, key4=Map, key6=1, name=numRecordsIn><>Count
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=taskmanager, key2=(.*), key3=(.*), key4=(.*), key5=([0-9]*), name=([a-z].*)><>Count
+    name: flink_taskmanager_$6
+    labels:
+      host: $1
```
Comment (brian-brazil) [inline example_configs/flink.yml]: what is this label exactly?

File: example_configs/flink.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=jobmanager, key2=(.*), name=([a-z].*)><>Value
+    name: flink_jobmanager_$3
+    labels:
+      ip: $1
+      job: $2
+
+  ## Task Manager ##
+  # Example: org.apache.flink.metrics<key0=debian-8, key1=taskmanager, key2=407401ecfef1f050d3dcc83af9cbfb74, key3=My_Job, key4=Map, key6=1, name=numRecordsIn><>Count
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=taskmanager, key2=(.*), key3=(.*), key4=(.*), key5=([0-9]*), name=([a-z].*)><>Count
+    name: flink_taskmanager_$6
```
Comment (brian-brazil) [inline example_configs/flink.yml]: these are counters, so should be suffixed with _total

File: example_configs/flink.yml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  - pattern: org.apache.flink.metrics<key0=(.*), key1=taskmanager, key2=(.*), key3=(.*), key4=(.*), key5=([0-9]*), name=([a-z].*)><>Value
+    name: flink_taskmanager_$6
+    labels:
+      host: $1
+      taskmanager_id: $2
+      job: $3
+      operator: $4
+      partition: $5
+
+  # Example: org.apache.flink.metrics<key0=debian-8, key1=taskmanager, key2=407401ecfef1f050d3dcc83af9cbfb74, key3=My_Job, key4=Sink, key5=1, key6=KafkaProducer, name=batch-size-avg><>Value
```
Comment (brian-brazil) [inline example_configs/flink.yml]: Are these the standard kafka metrics, just from a different mbean to normal?

File: example_configs/flink.yml (hunk 4 above)
Comment (mindprince) [inline example_configs/flink.yml]: Updated.

File: example_configs/flink.yml (hunk 2 above)
Comment (mindprince) [inline example_configs/flink.yml]: cardinality = equal to the number of taskmanagers running. Usually there's one taskmanager per node. When you restart the taskmanager (which happens rarely), the id changes.

File: example_configs/flink.yml (hunk 1 above)
Comment (mindprince) [inline example_configs/flink.yml]: Updated.

File: example_configs/flink.yml (hunk 2 above)
Comment (brian-brazil) [inline example_configs/flink.yml]: Is a taskmaster one process?

This sounds like a duplicate of the instance label.

File: example_configs/flink.yml (hunk 3 above)
Comment (mindprince) [inline example_configs/flink.yml]: key0 in taskmanager metrics is set to the hostname on which taskmanager is running.
So, on my dev environment it's set to `debian-8` (which is also the output of `hostname`)
On one of the prod nodes, it's set to the `hostname`.

Comment (brian-brazil) [inline example_configs/flink.yml]: That shouldn't be exposed then, as this is already covered by the instance label.

Comment (mindprince) [inline example_configs/flink.yml]: It's similar to the instance label. The instance label is usually like "host.example.com:PORT". This is just "host". I can remove it if you really don't want it.

File: example_configs/flink.yml (hunk 2 above)
Comment (mindprince) [inline example_configs/flink.yml]: Yeah, the taskmanager is one process. I would like to keep it, this helps in telling when the taskmanager was restarted.

File: example_configs/flink.yml (hunk 5 above)
Comment (mindprince) [inline example_configs/flink.yml]: When your flink job uses kafka sources or sinks, the kafka metrics are exposed like this.

File: example_configs/flink.yml (hunk 3 above)
Comment (brian-brazil) [inline example_configs/flink.yml]: It duplicates the role of the instance label, and it's up to the Prometheus user to decide what they want their instance label to be (which may not be the hostname).

Comment (mindprince) [inline example_configs/flink.yml]: Removed host label.

File: example_configs/flink.yml (hunk 2 above)
Comment (brian-brazil) [inline example_configs/flink.yml]: This should be removed then, it's not helping to distinguish the metrics.

Any time the same label is being applied to an entire target from the /metrics page is a smell, and it rarely makes sense to keep it.

Is `ip` also one of those?

`process_start_time_seconds` from the standard exports will cover you for restarts.

Comment (mindprince) [inline example_configs/flink.yml]: Done.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by thiszyq** [comment]
Hi, i want test the JMX_exporter in flink at local ubuntu. 
I haved start a prometheus server successed and start flink successed by **/bin/start-local.sh**. and i have download **jmx_prometheus_javaagent-0.9.jar** in my flink folder, and **flink-config.yml** is [example_configs/flink.yml](https://github.com/prometheus/jmx_exporter/blob/master/example_configs/flink.yml)
my prometheus.yml as follow:
```
global:
 scrape_interval: 10s
 evaluation_interval: 10s
scrape_configs:
 - job_name: 'flink'
   static_configs:
    - targets:
      - localhost:8001
```
what command should i input in bash so that i could get flink-metrics by wget http://localhost:8001/metrics 
Thank you in advance for any help you can provide.

**Comment by mindprince** [comment]
In your `flink-conf.yaml`, add
```
env.java.opts: -javaagent:/path/to/jmx_prometheus_javaagent-0.9.jar=7076:/path/to/example_configs/flink.yml
```
Then when you start the flink daemons:
The taskmanager metrics would be on http://taskmanager_host:7076/metrics
The jobmanager metrics would be on http://jobmanager_host:7076/metrics

**Comment by thiszyq** [comment]
thanks 

I had operated as your said ,then use *./bin/start-local.sh* to start flink daemon. but it not work, cant get metrics from http://jobmanager_host:port/metrics. ,also taskmanager-metrics.

but  could get jobmanager-metrics by add **$JVM_ARGS** in /bin/jobmanager.sh 

`export JVM_ARGS="$JVM_ARGS -Xms"$FLINK_JM_HEAP"m -Xmx"$FLINK_JM_HEAP"m -javaagent:$PWD/jmx_prometheus_javaagent-0.9.jar=9001:$PWD/flink.yml"`
@mindprince

**Comment by mindprince** [comment]
I haven't used local mode in Flink.

Basically, both the taskmanager and jobmanager JVM should know about the javaagent. If they are running on the same host, use different ports.

**Pull request merged**